sql/core/src/test/scala/org/apache/spark/sql/DataFrameRangeSuite.scala

```diff
@@ -127,4 +133,28 @@ class DataFrameRangeSuite extends QueryTest with SharedSQLContext {
       }
     }
   }
+
+  test("Cancelling stage in a query with Range.") {
+    val listener = new SparkListener {
+      override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+        Thread.sleep(100)
```

Comment:
Sorry that I didn't see `sparkContext.cancelStage(taskStart.stageId)`. How about this:

```Scala
class DataFrameRangeSuite {
  ...
  test("Cancelling stage in a query with Range.") {
    DataFrameRangeSuite.isTaskStarted = false
    val listener = new SparkListener {
      override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
        eventually(timeout(20.seconds)) {
          assert(DataFrameRangeSuite.isTaskStarted)
        }
        sparkContext.cancelStage(taskStart.stageId)
      }
    }
    sparkContext.addSparkListener(listener)
    for (codegen <- Seq(true, false)) {
      withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> codegen.toString()) {
        import testImplicits._
        val ex = intercept[SparkException] {
          spark.range(100000L).map { i =>
            DataFrameRangeSuite.isTaskStarted = true
            i
          }.crossJoin(spark.range(100000L))
            .toDF("a", "b").agg(sum("a"), sum("b")).collect()
        }
        ex.getCause() match {
          case null =>
            assert(ex.getMessage().contains("cancelled"))
          case cause: SparkException =>
            assert(cause.getMessage().contains("cancelled"))
          case cause: Throwable =>
            fail("Expected the casue to be SparkException, got " + cause.toString() + " instead.")
        }
      }
      eventually(timeout(20.seconds)) {
        assert(sparkContext.statusTracker.getExecutorInfos.map(_.numRunningTasks()).sum == 0)
      }
    }
  }
}

object DataFrameRangeSuite {
  @volatile var isTaskStarted = false
}
```